kotlin/show-location-history/src/main/java/com/esri/arcgisruntime/sample/showlocationhistory/MainActivity.kt

```diff
@@ -0,0 +1,180 @@
+    locationHistoryLineOverlay.renderer = SimpleRenderer(locationLineSymbol)
+
+    mapView.apply {
+      // set the map to a dark gray canvas basemap
+      map = ArcGISMap("https://www.arcgis.com/home/item.html?id=1970c1995b8f44749f4b9b6e81b5ba45")
+      // set the viewpoint
+      setViewpoint(Viewpoint(center, 7000.0))
+      // add the graphics overlays to the map view
+      graphicsOverlays.addAll(listOf(locationHistoryOverlay, locationHistoryLineOverlay))
+    }
+
+    // create a polyline builder to connect the location points
+    val polylineBuilder = PolylineBuilder(SpatialReference.create(102100))
+
+    // create a simulated location data source from json data with simulation parameters to set a consistent velocity
+    val simulatedLocationDataSource = SimulatedLocationDataSource().apply {
+      setLocations(
+        Polyline.fromJson(polylineJson) as Polyline,
+        SimulationParameters(Calendar.getInstance(), 30.0, 0.0, 0.0)
+      )
+    }
+
+    simulatedLocationDataSource.addLocationChangedListener { locationChangedEvent ->
+      // get the point from the location changed event
+      val nextPoint = locationChangedEvent.location.position
+      // add the new point to the polyline builder
+      polylineBuilder.addPoint(nextPoint)
+      // add the new point to the two graphics overlays and reset the line connecting the points
+      locationHistoryOverlay.graphics.add(Graphic(nextPoint))
+      locationHistoryLineOverlay.graphics.apply {
+        clear()
+        add(Graphic(polylineBuilder.toGeometry()))
+      }
+    }
+
+    // configure the map view's location display to follow the simulated location data source
+    mapView.locationDisplay.apply {
+      locationDataSource = simulatedLocationDataSource
+      autoPanMode = LocationDisplay.AutoPanMode.RECENTER
+      initialZoomScale = 7000.0
+    }
+
+    // start and stop the simulated location data source when the button is tapped
+    button.setOnClickListener {
+      if (simulatedLocationDataSource.isStarted) {
+        simulatedLocationDataSource.stop()
+        button.setImageResource(R.drawable.ic_my_location_white_24dp)
+      } else {
+        simulatedLocationDataSource.startAsync()
+        button.setImageResource(R.drawable.ic_navigation_white_24dp)
+      }
+    }
+
+    // make sure the floating action button doesn't obscure the attribution bar
+    mapView.addAttributionViewLayoutChangeListener { _, _, _, _, bottom, _, _, _, oldBottom ->
+      val layoutParams = button.layoutParams as CoordinatorLayout.LayoutParams
+      layoutParams.bottomMargin += bottom - oldBottom
+    }
+  }
+
+  override fun onResume() {
+    super.onResume()
+    mapView.resume()
+  }
+
+  override fun onPause() {
+    mapView.pause()
+    super.onPause()
+  }
+
+  override fun onDestroy() {
+    mapView.dispose()
+    super.onDestroy()
+  }
+}
+
+const val polylineJson = "{\"paths\":[[" +
```

Comment:
could this be moved to the strings resources xml?

Reply:
Yes it can, and I made that change, but it seems like the "correct" way to do this would be to have it as a raw file resource and then use a some kind of stream to read it as json. I've got it working with the string resources, but let me know if you think it's better to add the extra code.